framework/src/userobject/NodalNormalsUserObject.C

```diff
@@ -0,0 +1,181 @@
+  if (hasBoundary(node_boundary_ids, BoundaryRestrictable::ANY))
+  {
+    const dof_id_type & node_id = node->id();
+    Point normal_contribution(grad(0), grad(1), grad(2));
+    add(node_id, normal_contribution);
+  }
+}
+
+void
+NodalNormalsUserObject::add(dof_id_type node_id, const Point & contr) const
+{
+  const auto & it = _nodal_normals.find(node_id);
+  if (it == _nodal_normals.end())
+    _nodal_normals.insert(std::pair<dof_id_type, Point>(node_id, contr));
+  else
+    it->second += contr;
+}
+
+void
+NodalNormalsUserObject::zeroNormals() const
+{
+  for (auto & it : _nodal_normals)
+    it.second = Point(0, 0, 0);
+}
+
+void
+NodalNormalsUserObject::computeNormals() const
+{
+  for (auto & it : _nodal_normals)
+  {
+    Point n = it.second;
+    Real len = std::sqrt((n(0) * n(0)) + (n(1) * n(1)) + (n(2) * n(2)));
+    if (std::abs(len) >= 1e-13)
+    {
+      n /= len;
+      it.second = n;
+    }
+  }
+}
+
+void
+NodalNormalsUserObject::serialize(std::string & serialized_buffer) const
+{
+  std::ostringstream oss;
+
+  unsigned int size = _nodal_normals.size();
+  oss.write((char *)&size, sizeof(size));
+
+  Point zero_normal(0, 0, 0);
+  for (auto & it : _nodal_normals)
+  {
+    dof_id_type node_id = it.first;
+    storeHelper(oss, node_id, const_cast<NodalNormalsUserObject *>(this));
+    storeHelper(oss, it.second, const_cast<NodalNormalsUserObject *>(this));
+  }
+
+  // Populate the passed in string pointer with the string stream's buffer contents
+  serialized_buffer.assign(oss.str());
+}
+
+void
+NodalNormalsUserObject::deserialize(std::vector<std::string> & serialized_buffers) const
+{
+  // The input string stream used for deserialization
+  std::istringstream iss;
+
+  mooseAssert(serialized_buffers.size() == _app.n_processors(),
+              "Unexpected size of serialized_buffers: " << serialized_buffers.size());
+
+  for (auto rank = decltype(_app.n_processors())(0); rank < serialized_buffers.size(); ++rank)
+  {
+    if (rank == processor_id())
+      continue;
+
+    iss.str(serialized_buffers[rank]); // populate the stream with a new buffer
+    iss.clear();                       // reset the string stream state
+
+    // Load the communicated data into all of the other processors' slots
+
+    unsigned int size = 0;
+    iss.read((char *)&size, sizeof(size));
+
+    for (unsigned int i = 0; i < size; i++)
+    {
+      dof_id_type node_id;
+      loadHelper(iss, node_id, const_cast<NodalNormalsUserObject *>(this));
+
+      Point normal;
+      loadHelper(iss, normal, const_cast<NodalNormalsUserObject *>(this));
+
+      // merge the data we received from other procs
+      add(node_id, normal);
+    }
+  }
+}
+
+void
+NodalNormalsUserObject::communicate() const
+{
+  std::vector<std::string> send_buffers(1);
+  std::vector<std::string> recv_buffers;
+
+  recv_buffers.reserve(_app.n_processors());
+  serialize(send_buffers[0]);
+  comm().allgather_packed_range((void *)(nullptr),
```

Comment:
Not super scalable in terms of memory, but should be fine for awhile.

Reply:
This was a bitch to get right. The major problem is in the penetration system. Think of this case: 2 disconnected bodies each on a separate CPU. Nodal normals are computed on the right body. Then, the left body projects a node on the right body and needs the normals. So, I need to do this communication so that every CPU has all the normals. Luckily, we store normals only where they were requested (the old system has them in every node of the mesh).

Reply:
Also, for the record. We compute the normals typically just once per simulation. When adaptivity is on, then we need to recompute because the mesh changed. People can request computing the normals more often, but it is not typically needed.